source/extensions/filters/http/local_ratelimit/local_ratelimit.h

```diff
@@ -42,6 +42,19 @@ struct LocalRateLimitStats {
   ALL_LOCAL_RATE_LIMIT_STATS(GENERATE_COUNTER_STRUCT)
 };
 
+using LocalRateLimiterImplSharedPtr =
+    std::shared_ptr<Filters::Common::LocalRateLimit::LocalRateLimiterImpl>;
+
+class PerConnectionRateLimiter : public StreamInfo::FilterState::Object {
+public:
+  PerConnectionRateLimiter(LocalRateLimiterImplSharedPtr rl) : rate_limiter_(rl) {}
+  static const std::string& key();
+  LocalRateLimiterImplSharedPtr value() const { return rate_limiter_; }
+
+private:
+  LocalRateLimiterImplSharedPtr rate_limiter_;
```

Comment:
does this really need to be shared? Sounds like it could be unique or -- possibly better -- just an instance constructed when PerConnectionRateLimiter is created? `value()` would then just return a `LocalRateLimiterImpl&`.

Reply:
Seems like this can only be a `shared_ptr`. `PerConnectionRateLimiter` stores the state, i.e. an object of type `LocalRateLimiterImpl`,  inside it but since the latter does not have a copy constructor, you cannot pass in an already created instance nor can you create it within `PerConnectionRateLimiter` and return it since both of these would need a copy constructor. 
A `unique_ptr` wouldn't work for similar reasons since multiple, potentially concurrent, streams would be operating on a single instance of `LocalRateLimiterImpl` so transferring ownership of the unique_ptr out of `PerConnectionRateLimiter` and to each calling stream wouldn't make sense, if I'm understanding this correctly.